include/nbl/builtin/hlsl/bxdf/ndf.hlsl

```diff
@@ -0,0 +1,339 @@
+        scalar_type denom = params.n_or_a2 * params.NdotH2 * params.NdotH2;
+        return numbers::inv_pi<scalar_type> * nom / denom;
+    }
+
+    scalar_type operator()(SAnisotropicParams<scalar_type> params)
+    {
+        scalar_type nom = exp<scalar_type>(-(params.TdotH2 / params.ax2 + params.BdotH2 / params.ay2) / params.NdotH2);
+        scalar_type denom = params.ax * params.ay * params.NdotH2 * params.NdotH2;
+        return numbers::inv_pi<scalar_type> * nom / denom;
+    }
+};
+
+
+template<typename T NBL_PRIMARY_REQUIRES(is_scalar_v<T>)
+struct GGX
+{
+    using scalar_type = T;
+
+    // trowbridge-reitz
+    scalar_type operator()(SIsotropicParams<scalar_type> params)
+    {
+        scalar_type denom = params.NdotH2 * (params.n_or_a2 - 1.0) + 1.0;
+        return params.n_or_a2 * numbers::inv_pi<scalar_type> / (denom * denom);
+    }
+
+    scalar_type operator()(SAnisotropicParams<scalar_type> params)
+    {
+        scalar_type a2 = params.ax * params.ay;
+        scalar_type denom = params.TdotH2 / params.ax2 + params.BdotH2 / params.ay2 + params.NdotH2;
+        return numbers::inv_pi<scalar_type> / (a2 * denom * denom);
+    }
+
+    // burley
+    scalar_type operator()(SAnisotropicParams<scalar_type> params, scalar_type anisotropy)
+    {
+        scalar_type antiAniso = 1.0 - anisotropy;
+        scalar_type atab = params.ax * antiAniso;
+        scalar_type anisoTdotH = antiAniso * params.TdotH;
+        scalar_type anisoNdotH = antiAniso * params.NdotH;
+        scalar_type w2 = antiAniso/(params.BdotH * params.BdotH + anisoTdotH * anisoTdotH + anisoNdotH * anisoNdotH * params.ax);
+        return w2 * w2 * atab * numbers::inv_pi<scalar_type>;
+    }
+};
+
+// common
+namespace impl
+{
+template<class T, class U>
+struct is_ggx : bool_constant<
+    is_same<T, GGX<U> >::value
+> {};
+}
+
+template<class T> 
+struct is_ggx : impl::is_ggx<T, typename T::scalar_type> {};
+
+template<typename T>
+NBL_CONSTEXPR bool is_ggx_v = is_ggx<T>::value;
+
+
+enum MicrofacetTransformTypes : uint16_t
+{
+    REFLECT_BIT = 0b01,
+    REFRACT_BIT = 0b10,
+    REFLECT_REFRACT_BIT = 0b11
+};
+
+template<typename NDF, uint16_t reflect_refract>
+struct microfacet_to_light_measure_transform;
+
+
+template<typename NDF>
+struct microfacet_to_light_measure_transform<NDF,REFLECT_BIT>
+{
+    using this_t = microfacet_to_light_measure_transform<NDF,REFLECT_BIT>;
+    using scalar_type = typename NDF::scalar_type;
+
+    static this_t create(scalar_type NDFcos, scalar_type maxNdotV)
+    {
+        this_t retval;
+        retval.NDFcos = NDFcos;
+        if (is_ggx_v<NDF>)
+            retval.maxNdotL = maxNdotV;
+        else
+            retval.maxNdotV = maxNdotV;
+        return retval;
+    }
+
+    scalar_type operator()()
+    {
+        if (is_ggx_v<NDF>)
+            return NDFcos * maxNdotL;
+        else
+            return 0.25 * NDFcos / maxNdotV;
+    }
```

Comment:
write a comment that this computes the `max(NdotL,0)/(4*max(NdotV,0)*max(NdotL,0))` factor which transforms PDFs in the `f` in projected microfacet `f * NdotH` measure to projected light measure `f * NdotL`